src/VirtoCommerce.Platform.Core/Extensions/TypeExtensions.cs

```diff
@@ -67,5 +71,86 @@ private static string PrettyPrintRecursive(Type type, int depth)
                 : $"{nameParts[0]}<{string.Join(",", genericArguments.Select(t => PrettyPrintRecursive(t, depth + 1)))}>";
         }
 
+
+        /// <summary>
+        /// Check object clone immutability and independency:
+        /// Ensures the data in object clone equial to original.
+        /// Ensures no shared references between original and cloned objects (each object is a fully independent).
+        /// </summary>
+        /// <param name="original"></param>
+        /// <returns></returns>
+        public static async Task AssertCloneIndependency(this ICloneable original)
+        {
+            await Task.Run(() =>
+            {
+                var clone = original.Clone();
+                var sOriginal = JsonSerializer.Serialize((object)original, new JsonSerializerOptions() { WriteIndented = true });
+                var sClone = JsonSerializer.Serialize(clone, new JsonSerializerOptions() { WriteIndented = true });
+                if (!sOriginal.Equals(sClone)) // Ensure data in objects is equal
+                {
+                    throw new PlatformException(@$"Clone check failed: object and clone not equal.");
+                }
+                original.AssertNoSharedRefsWith(clone); // Ensure no shared references between objects (each object is a fully independent)
+            });
+        }
+
+        /// <summary>
+        /// Ensures no shared references with expected object (original and expected objects are fully independent from each other).
+        /// </summary>
+        /// <param name="original"></param>
+        /// <param name="expected"></param>
+        public static void AssertNoSharedRefsWith(this object original, object expected)
+        {
+            AssertNoSharedRefsWith(original, expected, new List<object>(), original.GetType().Name);
+        }
+
+        private static void AssertNoSharedRefsWith(object original, object expected, List<object> visited, string memberPath)
```

Comment:
maybe better would be narrow down this method responsibility?
Rename this method to something like this  IEnumerable<object> GetAllObjectPropertiesRecurively(obj) where return all  found properties values via `yield return  propValue` expression.